exporter/exporterhelper/queued_retry.go

```diff
@@ -124,29 +136,123 @@ func createSampledLogger(logger *zap.Logger) *zap.Logger {
 	return logger.WithOptions(opts)
 }
 
-func newQueuedRetrySender(fullName string, qCfg QueueSettings, rCfg RetrySettings, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
+func newQueuedRetrySender(id config.ComponentID, qCfg QueueSettings, rCfg RetrySettings, reqUnmarshaler requestUnmarshaler, nextSender requestSender, logger *zap.Logger) *queuedRetrySender {
 	retryStopCh := make(chan struct{})
 	sampledLogger := createSampledLogger(logger)
-	traceAttr := trace.StringAttribute(obsreport.ExporterKey, fullName)
-	return &queuedRetrySender{
-		fullName: fullName,
-		cfg:      qCfg,
-		consumerSender: &retrySender{
-			traceAttribute: traceAttr,
-			cfg:            rCfg,
-			nextSender:     nextSender,
-			stopCh:         retryStopCh,
-			logger:         sampledLogger,
-		},
-		queue:           queue.NewBoundedQueue(qCfg.QueueSize, func(item interface{}) {}),
-		retryStopCh:     retryStopCh,
-		traceAttributes: []trace.Attribute{traceAttr},
-		logger:          sampledLogger,
+	traceAttr := trace.StringAttribute(obsreport.ExporterKey, id.String())
+
+	qrs := &queuedRetrySender{
+		id:                 id,
+		cfg:                qCfg,
+		retryStopCh:        retryStopCh,
+		traceAttributes:    []trace.Attribute{traceAttr},
+		logger:             sampledLogger,
+		requestUnmarshaler: reqUnmarshaler,
+	}
+
+	qrs.consumerSender = &retrySender{
+		traceAttribute: traceAttr,
+		cfg:            rCfg,
+		nextSender:     nextSender,
+		stopCh:         retryStopCh,
+		logger:         sampledLogger,
+		// Following three functions are provided in such way because they actually depend on queuedRetrySender
+		onSuccess:          qrs.onSuccess,
+		onTemporaryFailure: qrs.onTemporaryFailure,
+		onPermanentFailure: qrs.onPermanentFailure,
 	}
+
+	return qrs
+}
+
+func getStorageClient(ctx context.Context, host component.Host, id config.ComponentID) (*storage.Client, error) {
+	var storageExtension storage.Extension
+	for _, ext := range host.GetExtensions() {
+		if se, ok := ext.(storage.Extension); ok {
+			if storageExtension != nil {
+				return nil, errMultipleStorageClients
+			}
+			storageExtension = se
+		}
+	}
+
+	if storageExtension == nil {
+		return nil, errNoStorageClient
+	}
+
+	client, err := storageExtension.GetClient(ctx, component.KindReceiver, id)
```

Comment:
Based on the context, I think you should probably use `component.KindExporter`. (This only affects the name of the storage file.)